// node-pg-migrate/dist/operations/viewsMaterialized.js
"use strict";

const _require = require("../utils"),
      template = _require.template,
      quote = _require.quote,
      formatLines = _require.formatLines;

const dataClause = data => data !== undefined ? ` WITH${data ? "" : " NO"} DATA` : "";

const storageParameterStr = storageParameters => key => {
  const value = storageParameters[key] === true ? "" : ` = ${storageParameters[key]}`;
  return `${key}${value}`;
};

function dropMaterializedView(viewName, {
  ifExists,
  cascade
} = {}) {
  const ifExistsStr = ifExists ? " IF EXISTS" : "";
  const cascadeStr = cascade ? " CASCADE" : "";
  return template`DROP MATERIALIZED VIEW${ifExistsStr} "${viewName}"${cascadeStr};`;
}

function createMaterializedView(viewName, options, definition) {
  const ifNotExists = options.ifNotExists,
        _options$columns = options.columns,
        columns = _options$columns === void 0 ? [] : _options$columns,
        tablespace = options.tablespace,
        _options$storageParam = options.storageParameters,
        storageParameters = _options$storageParam === void 0 ? {} : _options$storageParam,
        data = options.data; // prettier-ignore

  const columnNames = quote(Array.isArray(columns) ? columns : [columns]).join(", ");
  const withOptions = Object.keys(storageParameters).map(storageParameterStr(storageParameters)).join(", ");
  const ifNotExistsStr = ifNotExists ? " IF NOT EXISTS" : "";
  const columnsStr = columnNames ? `(${columnNames})` : "";
  const withOptionsStr = withOptions ? ` WITH (${withOptions})` : "";
  const tablespaceStr = tablespace ? `TABLESPACE ${tablespace}` : "";
  const dataStr = dataClause(data);
  return template`CREATE MATERIALIZED VIEW${ifNotExistsStr} "${viewName}"${columnsStr}${withOptionsStr}${tablespaceStr} AS ${definition}${dataStr};`;
}

function alterMaterializedView(viewName, options) {
  const cluster = options.cluster,
        extension = options.extension,
        _options$storageParam2 = options.storageParameters,
        storageParameters = _options$storageParam2 === void 0 ? {} : _options$storageParam2;
  const clauses = [];

  if (cluster !== undefined) {
    if (cluster) {
      clauses.push(`CLUSTER ON "${cluster}"`);
    } else {
      clauses.push(`SET WITHOUT CLUSTER`);
    }
  }

  if (extension) {
    clauses.push(`DEPENDS ON EXTENSION "${extension}"`);
  }

  const withOptions = Object.keys(storageParameters).filter(key => storageParameters[key]).map(storageParameterStr(storageParameters)).join(", ");

  if (withOptions) {
    clauses.push(`SET (${withOptions})`);
  }

  const resetOptions = Object.keys(storageParameters).filter(key => !storageParameters[key]).join(", ");

  if (resetOptions) {
    clauses.push(`RESET (${resetOptions})`);
  }

  const clausesStr = formatLines(clauses);
  return template`ALTER MATERIALIZED VIEW "${viewName}"\n${clausesStr};`;
}

function renameMaterializedView(viewName, newViewName) {
  return template`ALTER MATERIALIZED VIEW "${viewName}" RENAME TO "${newViewName}";`;
}

const undoRename = (viewName, newViewName) => renameMaterializedView(newViewName, viewName);

function renameMaterializedViewColumn(viewName, columnName, newColumnName) {
  return template`ALTER MATERIALIZED VIEW "${viewName}" RENAME COLUMN ${columnName} TO "${newColumnName}";`;
}

const undoRenameColumn = (viewName, columnName, newColumnName) => renameMaterializedViewColumn(viewName, newColumnName, columnName);

function refreshMaterializedView(viewName, {
  concurrently,
  data
} = {}) {
  const concurrentlyStr = concurrently ? " CONCURRENTLY" : "";
  const dataStr = dataClause(data);
  return template`REFRESH MATERIALIZED VIEW${concurrentlyStr} "${viewName}"${dataStr};`;
}

createMaterializedView.reverse = dropMaterializedView;
renameMaterializedView.reverse = undoRename;
renameMaterializedViewColumn.reverse = undoRenameColumn;
refreshMaterializedView.reverse = refreshMaterializedView;
module.exports = {
  createMaterializedView,
  dropMaterializedView,
  alterMaterializedView,
  renameMaterializedView,
  renameMaterializedViewColumn,
  refreshMaterializedView
};
